### PR TITLE
Updated ByteReader to allow white space and comments content after --- when splitting YAML documents

### LINKED ISSUE #3876: Invalid build output if resource separator has trailing spaces
If a file has multiple resources separated by `---` and the separator is followed by trailing spaces then `kustomize build` silently generates an invalid output.

**Files that can reproduce the issue**

**kustomization.yaml**:
```yaml
cat <<EOF >./kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- install.yaml
EOF
```
**install.yaml**:
```yaml
cat <<EOF >./install.yaml
apiVersion: v1
kind: Namespace
metadata:
  name: test
---    #trailing spaces here
apiVersion: v1
kind: Service
metadata:
  labels:
    app: test
  name: test
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: test
  type: ClusterIP
EOF
```

**Expected output**

An error message indicating the file {full-file-name} is invalid due to trailing spaces in line {line-number}. Alternatively if this is allowed the output should be:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: test
  name: test
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: test
  type: ClusterIP
```

**Actual output**

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: test

```

**Kustomize version**

```
{Version:kustomize/v4.1.2 GitCommit:a5914abad89e0b18129eaf1acc784f9fe7d21439 BuildDate:2021-04-15T20:38:06Z GoOs:linux GoArch:amd64}
```
**Platform**

Linux

**Additional context**

### PR DESCRIPTION
**Pull request by brianpursley**
Instead of splitting on `\n---\n`, split on `\n---.*\n` regular expression.

Allow (and discard) any white space or comments,  and return an error if other content exists on the same line as `---`

Fixes #3876

### CODE REVIEW

File: kyaml/kio/byteio_reader.go [hunk 1]
```diff
@@ -133,6 +134,8 @@ type ByteReader struct {
 
 var _ Reader = &ByteReader{}
 
+var yamlSeparatorRegexp = regexp.MustCompile(`\n---.*\n`)
```
Comment (natasha41575) [inline kyaml/kio/byteio_reader.go:137]: what's the motivation for declaring this outside of the function?

File: kyaml/kio/byteio_reader.go [hunk 2]
```diff
@@ -144,8 +147,8 @@ func (r *ByteReader) Read() ([]*yaml.RNode, error) {
 		return nil, errors.Wrap(err)
 	}
 
-	// replace the ending \r\n (line ending used in windows) with \n and then separate by \n---\n
-	values := strings.Split(strings.ReplaceAll(input.String(), "\r\n", "\n"), "\n---\n")
+	// replace the ending \r\n (line ending used in windows) with \n and then split on the YAML document separator, ---
+	values := yamlSeparatorRegexp.Split(strings.ReplaceAll(input.String(), "\r\n", "\n"), -1)
```
Comment (natasha41575) [inline kyaml/kio/byteio_reader.go:151]: let's add a comment for why we are doing this

File: kyaml/kio/byteio_reader_test.go [hunk 3]
```diff
@@ -372,6 +372,27 @@ metadata:
...
 		},
+
+		//
+		//
+		//
+		{
+			name: "any_content_after_document_separator_should_be_ignored",
+			input: `
+a: b
+--- foo
```
Comment (natasha41575) [inline kyaml/kio/byteio_reader_test.go:383]: would it make more sense to throw an error in this case or just discard it like we would a comment?

let's also add a test with a comment, `--- #foo`, as a regression test

@KnVerey

File: kyaml/kio/byteio_reader.go (hunk 1 above)
Comment (brianpursley) [reply kyaml/kio/byteio_reader.go:137]: It is to avoid compiling the same regexp multiple times if `Read()` is invoked multiple times.

I've seen it done this way frequently in k/k and have taken this approach myself before too, but to be honest I don't know how much of an impact it would really have on things.

I'll move it inside the function and we can deal with it later if it becomes an issue.

File: kyaml/kio/byteio_reader_test.go (hunk 3 above)
Comment (brianpursley) [reply kyaml/kio/byteio_reader_test.go:383]: I can add a couple more tests (one for whitespace and another for comments).

As far as detecting invalid content after `---` I chose not to do that in this PR because that adds quite a bit more complexity.  Instead of just splitting, it would have to find and detect the invalid situations as well.  I can do it but was trying to keep this PR as simple as possible.

---

Also, interestingly, I just now discovered that k/k doesn't handle the following as I thought it would:
```
kubectl --dry-run=server apply -f - << EOF
apiVersion: v1
kind: Namespace
metadata:
  name: foo
--- # something
apiVersion: v1
kind: Namespace
metadata:
  name: bar
EOF
```

It doesn't fail, but it only creates foo, not bar.

Output:
```
namespace/foo created (server dry run)
```

It does seem to allow whitespace after `---` though.

Comment (natasha41575) [reply kyaml/kio/byteio_reader_test.go:425]: > Instead of just splitting, it would have to find and detect the invalid situations as well. I can do it but was trying to keep this PR as simple as possible.

That makes sense we can keep it simple. Could you add a TODO comment, file a followup issue, and link the issue in the TODO? 

> It doesn't fail, but it only creates foo, not bar.

Like I said on the issue, that sounds like a bug they need to fix. It's good that we are trying to address it in kustomize.

Comment (natasha41575) [reply kyaml/kio/byteio_reader_test.go:425]: And thanks for the extra tests, they look great!

Comment (brianpursley) [reply kyaml/kio/byteio_reader_test.go:425]: @natasha41575 I just pushed another commit with a version that detects anything other than white space and comments and returns an error.

My original intent was to go for a simple change, but if you are interested in it being more robust in checking the separator lines, take a look at what I just pushed.

Sorry for getting the communication lines crossed up.  Please take a look and if you want to opt for the original, simple version I can revert back to that. :smile:

Comment (natasha41575) [reply kyaml/kio/byteio_reader_test.go:425]: awesome, thank you! Yes I prefer robust checking so the error check is great.

**Comment by natasha41575** [comment]
/lgtm
I'll leave this PR open for a little bit in case others want to take a look.
